prow/apis/prowjobs/v1/types_test.go

```diff
@@ -0,0 +1,374 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1
+
+import (
+	"reflect"
+	"testing"
+	"time"
+)
+
+func TestDecorationDefaulting(t *testing.T) {
+	truth := true
+	lies := false
+	defaults := &DecorationConfig{
+		Timeout:     1 * time.Minute,
+		GracePeriod: 10 * time.Second,
+		UtilityImages: &UtilityImages{
+			CloneRefs:  "clonerefs",
+			InitUpload: "iniupload",
+			Entrypoint: "entrypoint",
+			Sidecar:    "sidecar",
+		},
+		GCSConfiguration: &GCSConfiguration{
+			Bucket:       "bucket",
+			PathPrefix:   "prefix",
+			PathStrategy: PathStrategyLegacy,
+			DefaultOrg:   "org",
+			DefaultRepo:  "repo",
+		},
+		GCSCredentialsSecret: "secretName",
+		SSHKeySecrets:        []string{"first", "second"},
+		SSHHostFingerprints:  []string{"primero", "segundo"},
+		SkipCloning:          &truth,
+	}
+
+	var testCases = []struct {
+		name     string
+		provided *DecorationConfig
+		expected *DecorationConfig
+	}{
+		{
+			name:     "nothing provided",
+			provided: &DecorationConfig{},
+			expected: &DecorationConfig{
+				Timeout:     1 * time.Minute,
+				GracePeriod: 10 * time.Second,
+				UtilityImages: &UtilityImages{
+					CloneRefs:  "clonerefs",
+					InitUpload: "iniupload",
+					Entrypoint: "entrypoint",
+					Sidecar:    "sidecar",
+				},
+				GCSConfiguration: &GCSConfiguration{
+					Bucket:       "bucket",
+					PathPrefix:   "prefix",
+					PathStrategy: PathStrategyLegacy,
+					DefaultOrg:   "org",
+					DefaultRepo:  "repo",
+				},
+				GCSCredentialsSecret: "secretName",
+				SSHKeySecrets:        []string{"first", "second"},
+				SSHHostFingerprints:  []string{"primero", "segundo"},
+				SkipCloning:          &truth,
+			},
+		},
+		{
+			name: "timeout provided",
+			provided: &DecorationConfig{
+				Timeout: 10 * time.Minute,
+			},
+			expected: &DecorationConfig{
```

Comment:
Contrast this with:

```golang
expected: func(left, right *DecorationConfig) *DecorationConfig {
  out := *left
  out.Timeout = right.Timeout
  return &out
}
```

or if you're careful to not pass the defaults reference:

```golang
expected: func(left, right *DecorationConfig) *DecorationConfig {
  left.Timeout = right.Timeout
  return left
}
```

This is much easier to read because it hides all the nonsense we don't care about